devilspie.1
.TH devilspie 1
.SH NAME
devilspie \- perform actions on windows as they are created

.SH SYNOPSIS
.B devilspie
.I [\fIOPTION\fR]...
.I "configuration-files..."

.SH "DESCRIPTION"
.BR devilspie
is a program which can detect windows as they are created, and perform
actions on them if they match as set of criteria.

The configuration files are a set of s-expressions.  For documentation and
examples, see /usr/share/doc/devilspie/.

Command-lines options:
.TP
\fB\-a\fR \fB\-\-apply\-to\-existing\fR
Apply to all existing windows instead of just new windows.
.TP
\fB\-d\fR \fB\-\-debug\fR
Output debug information
.TP
\fB\-l\fR \fB\-\-list\-symbols\fR
List available symbols supported in input files.
.TP
\fB\-s\fR \fIsymname\fR \fB\-\-symbol\-doc\fR=\fIsymname\fR
 Get detailed help on symbol \fIsymname\fR ('all' for all symbols).

.SH "ALLOWED SYMBOLS"
In the input files, the following symbols are known by \fIdevilspie\fR:
."SYMBOL HELP BEGINS HERE
.TP
\fBis\fR
String equality, (is a b) means a is the same as b.
.TP
\fBcontains\fR
Substring, (contains haystack needle) means haystack contains needle.
.TP
\fBmatches\fR
Regexp matches, (matches str pattern) means the regexp pattern matches str.
.TP
\fBwindow_name\fR
Return the title of the current window (String).
.TP
\fBwindow_role\fR
Return the role (as determined by the WM_WINDOW_ROLE hint) of the current window (String).
.TP
\fBwindow_class\fR
Return the class of the current window (String).
.TP
\fBwindow_xid\fR
Return the X11 window id of the current window (Integer).
.TP
\fBapplication_name\fR
Return the application name (as determined by libwnck) of the current window (String).
.TP
\fBwindow_property\fR
Returns the given property of the window, e.g. pass "_NET_WM_STATE" (String).
.TP
\fBwindow_workspace\fR
Returns the workspace a window is on (Integer).
.TP
\fBdebug\fR
Debugging function, outputs the current window's title, name, role and geometry (Returns TRUE).
.TP
\fBprint\fR
Print args (without trailing \n, returns boolean).
.TP
\fBprintln\fR
Print args (with trailing \n, returns boolean).
.TP
\fBstr\fR
Transform parameters into strings and concat them.
.TP
\fBhex\fR
Transform the integer parameter into an unsigned hexadecimal string (with 0x prefix).
.TP
\fBgeometry\fR
Set position + size (as tring) of current window (returns boolean).
.TP
\fBfullscreen\fR
Make the current window fullscreen (returns TRUE).
.TP
\fBfocus\fR
Focus the current window (returns TRUE).
.TP
\fBcenter\fR
Center position of current window (returns boolean).
.TP
\fBmaximize\fR
Maximise the current window (returns TRUE).
.TP
\fBmaximize_vertically\fR
Maximise vertically the current window (returns TRUE).
.TP
\fBmaximize_horizontally\fR
Maximise horizontally the current window (returns TRUE).
.TP
\fBunmaximize\fR
Un-maximise the current window (returns TRUE).
.TP
\fBminimize\fR
Minimise the current window (returns TRUE).
.TP
\fBunminimize\fR
Un-minimise the current window (returns TRUE).
.TP
\fBshade\fR
Shade ("roll up") the current window (returns TRUE).
.TP
\fBunshade\fR
Un-shade ("roll down") the current window (returns TRUE).
.TP
\fBclose\fR
Close the current window (returns TRUE).
.TP
\fBpin\fR
Pin the current window to all workspaces (returns TRUE).
.TP
\fBunpin\fR
Unpin the current window from all workspaces (returns TRUE).
.TP
\fBstick\fR
Make the current window stick to all viewports (returns TRUE).
.TP
\fBunstick\fR
Unstick the window from viewports (returns TRUE).
.TP
\fBset_workspace\fR
Move the window to a specific workspace number, counting from 1 (returns boolean).
.TP
\fBset_viewport\fR
Move the window to a specific viewport number, counting from 1 (returns boolean).
.TP
\fBskip_pager\fR
Remove the current window from the window list (returns TRUE).
.TP
\fBskip_tasklist\fR
Remove the current window from the pager (returns TRUE).
.TP
\fBabove\fR
Set the current window to be above all normal windows (returns TRUE).
.TP
\fBbelow\fR
Set the current window to be below all normal windows (returns TRUE).
.TP
\fBdecorate\fR
Add the window manager decorations to the current window (returns boolean).
.TP
\fBundecorate\fR
Remove the window manager decorations from the current window (returns boolean).
.TP
\fBwintype\fR
Set the window type of the current window (returns boolean). Accepted values are: normal, dialog, menu, toolbar, splashscreen, utility, dock, desktop.
.TP
\fBopacity\fR
Change the opacity level (as integer in 0..100) of the current window (returns boolean).
.TP
\fBspawn_async\fR
Execute a command in the background (returns boolean). Command is given as a single string, or as a series of strings (similar to execl).
.TP
\fBspawn_sync\fR
Execute a command in the foreground (returns command output as string, or FALSE on error). Command is given as a single string, or as a series of strings (similar to execl).
."SYMBOL HELP ENDS HERE

.SH BUGS
This manpage is almost useless, as it was thrown together on a train.

.SH AUTHOR
This manual page was written by Ross Burton <ross@burtonini.com>.


NEWS
Devil's Pie "A Love Supreme" 0.21
=================================

* Add change_workspace action (#453464, thanks Adam Więckowski)
* Add decorate action (#449634, thanks Galkin Vasily)
* Add quit action (thanks Leon Zhang)
* Handle wnck_screen_get_active_workspace() failing (#343546)


Devil's Pie "Wearing That Dress" 0.20.2
=======================================

* Fix window_workspace (Andrew Yates)


Devil's Pie "Wearing That Dress" 0.20.1
=======================================

* Fix parsing (Lars Damerow)
* Fix test suite


Devil's Pie "Wearing That Dress" 0.20
=====================================

* Handle multiple expressions in a file (Lars Damerow)
* Add spawn_sync and spawn_async actions to start processes (David Decotigny)
* Add println, str, hex, and expand print (DD)
* Add window_xid matcher (DD)


Devil's Pie "Sell Sell Sell" 0.19
================================

* Don't use wnck_window_set_geometry (#381233)


Devil's Pie "Simmer To A Sizzle Like The Days Of Old" 0.18
==========================================================

* Add window_property matcher (Nigel Tao)
* Add set_viewport action (James Willcox)
* Add opacity action (Pavel Palat)
* Add sticky action (Steve Leung)
* Add unmaximise action (Alex Menk)
* Add unminimise action (Richard Neill)
* Fix negative offsets in geometry action (Vincent Ho, Midred)


Devil's Pie "Out Of Gas, Still Burning" 0.17.1
==============================================

* Depend on GLib 2.9.1 onwards


Devil's Pie "Stay With Me While We Grow Old" 0.17
=================================================

* Fix compile with GLib 2.10.


Devil's Pie "Love To Hate" 0.16
===============================

* Fix undecorate action (Mikachu, RB)


Devil's Pie "Shogun's Decapitator" 0.15
=======================================

* New matcher: window class (Jonas Linde)
* New actions: focus (JL), center (JL), print (Ryan Koppenhaver), close (RK).
* Fix test suite (RK)
* Don't segfault on invalid s-exp (Christian Krause)
* Never return NULL from a matcher
* Wrap X call when removing decorations with error handling


Devil's Pie "Lazy Release" 0.14
===============================

* Add Set Geometry (Guido Boehm)
* Fix Set Window Type (Stefan van der Haven)
* Make Pin work as expected (Jean-Yves Lefort)
* Append new expressions, not prepend (Lars Damerow)
* Fix typos in README (Larry Virden)


Devil's Pie "Wind The Frog" 0.13
================================

* Total rewrite.
* No, really.  A complete rewrite, no file left untouched.  Many thanks to Pixar
  Animation Studios for sponsoring this work.
* Use s-expressions instead of XML as the configuration file format, allowing
  matches to be combined in new and exciting ways with (and) (or) and (not)
  operators.
* Read configuration from both the users home directory and /etc, allowing
  centralised configuration.
* Add the beginnings of a test suite.


Devil's Pie "Trying To Make Their Paper" 0.12
=============================================

* Quit if no flurbs were loaded
* Use GOption
* Major code cleanup


Devil's Pie "Used To Love You" 0.11
===================================

* Add an action to shade windows by Anonymous.  I didn't write it but I've lost
  the mail the patch came in...
* Fix the window type action
* Fix the execute action
* Check for xsltproc in configure, as it's required


Devil's Pie "Can't Join Them? Beat Them" 0.10
=============================================

* Re-implement the required private wnck functions, now builds with G2.10
  (Crispin Flowerday rocks)
* Add min/max hints to the opacity action property
* Display the min/max hints in the generated documentation.


Devil's Pie "Missing Gifts" 0.9
===============================

* Check for libwnck 2.9.92 and pass timestamps if required
* Handle properties containing doubles, so the opacity action works
* Add CONSTRUCT flag to properties (Jean-Yves Lefort)
* Start a private copy of various private wnck functions we need


Devil's Pie "Mystery Boxes" 0.8
===============================

* Bring back --apply-to-existing, so you can work on existing windows
* Add maximized_horizontally and _vertically to the resize action (John Russell)
* Add an execute action (John Russell)
* Add an opacity setting action (Guido Boehm)
* Generate decent matcher/action documentation at build-time from the source
* Check that regular expressions parsed correctly, fixing a common bug
* Less compile warnings


Devil's Pie "Stumpy" 0.7
===============================

* Set the Motif window hint when setting the decoration boolean (Vaclav Lorenc)


Devil's Pie "Salmon Sunset" 0.6
===============================

* Fix the fullscreen action.  Sadly this bug has been in bugs.debian.net for 303
  days.


Devil's Pie "Right Where I Want You" 0.5
========================================

* Add an action to set the geometry of a window, by Henrik Brix Andersen.


Devil's Pie "Nothing To See" 0.4.1
==================================

* Don't ship the debian/ folder.


Devil's Pie "Derren Brown Is A Witch" 0.4
=========================================

* Add an action to set the window type atom (Olivier Andrieu)
* Add an action to set the decorate window hint in OpenBox (Vaclav Lorenc)
* Watch all screens for events, instead of just the default


Devil's Pie "David Blaine has smelly socks" 0.3.1
=================================================

* Don't crash when a window role is null in various places (Rob Mayoff)
* Don't crash on unknown properties (RM)
* Being very slow at applying patches (me)


Devil's Pie "Floating Skulls" 0.3
=================================

* Vastly improved documentation (Gaetan de Menten)
* Add a "fullscreen" property (requires libwnck 2.3.something)
* Added a RPM .spec file (me, Michael Raab, Lars R. Damerow)
* Many bug fixes (Patrick Aussems)

Translations:
* de (Christian Neumair)


Devil's Pie "Long Time Coming" 0.2.4
====================================


src/devilspie.c
/*
 * Copyright (C) 2005 Ross Burton <ross@burtonini.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdlib.h>
#include <glib.h>
#include <glib/gi18n.h>
#include <libwnck/libwnck.h>
#include "e-sexp.h"
#include "devilspie.h"
#include "parser.h"

/* Global state */

/**
 * The list of s-expressions to evaluate.
 */
GList *sexps = NULL;

/**
 * If we are in debug mode.
 */
gboolean debug = FALSE;

/**
 * Matching context
 */
Context context = {NULL,};

/* Private state */

/**
 * If we apply to existing windows or not.
 */
static gboolean apply_to_existing = FALSE;

/**
 * List of configuration files to open, or NULL if to open the default files.
 */
static char **files = NULL;

/**
 * The main loop.
 */
GMainLoop *loop = NULL;

/**
 * Evaluate a s-expression.
 */
static void run_sexp(ESExp * sexp, WnckWindow *window) {
  context.window = window;
  /* We really don't care what the top-level s-expressions evaluate to */
  e_sexp_result_free (sexp, e_sexp_eval(sexp));
}

/**
 * This callback is called whenever a window is opened on a screen.
 */
static void window_opened_cb(WnckScreen *screen, WnckWindow *window) {
  g_list_foreach(sexps, (GFunc)run_sexp, window);
  return;
}

/**
 * Connect to every screen on this display and watch for new windows.
 */
static void init_screens(void) {
  int i;
  const int num_screens = gdk_display_get_n_screens (gdk_display_get_default());
  for (i = 0 ; i < num_screens; ++i) {
    WnckScreen *screen = wnck_screen_get (i);
    /* Connect a callback to the window opened event in libwnck */
    g_signal_connect (screen, "window_opened", (GCallback)window_opened_cb, NULL);
    /* TODO: broken at the moment due to wnck change */
    //if (apply_to_existing) wnck_screen_force_update (screen);
  }
}

/*
 * Dedicated to Vicky.
 */
int main(int argc, char **argv) {
  static const GOptionEntry options[] = {
    { "apply-to-existing", 'a', 0, G_OPTION_ARG_NONE, &apply_to_existing, N_("Apply to all existing windows instead of just new windows."), NULL },
    { "debug", 'd', 0, G_OPTION_ARG_NONE, &debug, N_("Output debug information"), NULL },
    { G_OPTION_REMAINING, '\0', 0, G_OPTION_ARG_FILENAME_ARRAY, &files, N_("Configuration files to use"), NULL },
    { NULL }
  };

  GError *error = NULL;
  GOptionContext *context;

  /* Initialise i18n */
  bindtextdomain (GETTEXT_PACKAGE, LOCALEDIR);
  bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
  textdomain (GETTEXT_PACKAGE);

  /* Initialise GDK */
  gdk_init(&argc, &argv);

  /* Parse the arguments */
  context = g_option_context_new ("- Devil's Pie " VERSION);
  g_option_context_add_main_entries (context, options, GETTEXT_PACKAGE);
  g_option_context_parse (context, &argc, &argv, &error);

  if (debug) g_printerr(_("Devil's Pie %s starting...\n"), VERSION);

  /* If there were files specified use those, otherwise load the default configuration */
  if (files) {
    while (*files) {
      GList *s;
      s = load_configuration_file (*files++);
      if (s) sexps = g_list_concat (sexps, s);
    }
  } else {
    load_configuration ();
  }

  if (debug) g_printerr("%d s-expressions loaded.\n", g_list_length(sexps));

  if (g_list_length (sexps) == 0) {
    g_printerr(_("No s-expressions loaded, quiting\n"));
    return 1;
  }

  /* Connect to every screen */
  init_screens ();

  /* Go go go! */
  loop = g_main_loop_new (NULL, TRUE);
  g_main_loop_run (loop);

  return 0;
}
